ads/google/a4a/utils.js

```diff
@@ -124,116 +128,151 @@ export function isReportingEnabled(ampElement) {
-      Object.keys(containerTypeSet).join()});
-    if (isCanary(win)) {
-      // The semantics here are:
-      //   0: production branch (this is never actually sent)
-      //   1: control branch (this is not yet supported, so is never sent)
-      //   2: canary branch
-      queryParams.push({name: 'art', value: '2'});
-    }
-    let eids = adElement.getAttribute('data-experiment-id');
-    if (opt_experimentIds) {
-      eids = mergeExperimentIds(opt_experimentIds, eids);
-    }
-    const allQueryParams = queryParams.concat(
-      [
-        {
-          name: 'is_amp',
-          value: AmpAdImplementation.AMP_AD_XHR_TO_IFRAME_OR_AMP,
-        },
-        {name: 'amp_v', value: '$internalRuntimeVersion$'},
-        {name: 'd_imp', value: '1'},
-        {name: 'dt', value: startTime},
-        {name: 'ifi', value: slotNumber},
-        {name: 'adf', value: domFingerprint(adElement)},
-        {name: 'c', value: getCorrelator(win, clientId)},
-        {name: 'output', value: 'html'},
-        {name: 'nhd', value: iframeDepth},
-        {name: 'iu', value: adElement.getAttribute('data-ad-slot')},
-        {name: 'eid', value: eids},
-        {name: 'biw', value: viewportRect.width},
-        {name: 'bih', value: viewportRect.height},
-        {name: 'adx', value: slotRect.left},
-        {name: 'ady', value: slotRect.top},
-        {name: 'u_aw', value: screen ? screen.availWidth : null},
-        {name: 'u_ah', value: screen ? screen.availHeight : null},
-        {name: 'u_cd', value: screen ? screen.colorDepth : null},
-        {name: 'u_w', value: screen ? screen.width : null},
-        {name: 'u_h', value: screen ? screen.height : null},
-        {name: 'u_tz', value: -new Date().getTimezoneOffset()},
-        {name: 'u_his', value: getHistoryLength(win)},
-        {name: 'oid', value: '2'},
-        {name: 'brdim', value: additionalDimensions(win, viewportSize)},
-        {name: 'isw', value: viewportSize.width},
-        {name: 'ish', value: viewportSize.height},
-        {name: 'pfx', value: pfx},
-        {name: 'rc', value: a4a.fromResumeCallback ? 1 : null},
-      ],
-      unboundedQueryParams,
-      [
-        {name: 'url', value: documentInfo.canonicalUrl},
-        {name: 'top', value: iframeDepth ? topWindowUrlOrDomain(win) : null},
-        {
-          name: 'loc',
-          value: win.location.href == documentInfo.canonicalUrl ?
-            null : win.location.href,
-        },
-        {name: 'ref', value: referrer},
-      ]
-    );
-    const url = buildUrl(baseUrl, allQueryParams, MAX_URL_LENGTH - 10,
-                         {name: 'trunc', value: '1'});
-    return url + '&dtd=' + elapsedTimeWithCeiling(Date.now(), startTime);
-  }));
+  const blockLevelParameters = googleBlockParameters(a4a, opt_experimentIds);
+  return googlePageParameters(a4a.win, a4a.getAmpDoc(), startTime)
+    .then(pageLevelParameters => {
+      Object.assign(parameters, blockLevelParameters);
```

Comment:
I suppose that's fair but given the mutated object is constructed within this function and never exposed, I see no reason not to use it as opposed to creating a copy.